### PR TITLE
Automatically group sub-commands ##newshell

### PR DESCRIPTION
**Pull request by ret2libc**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

Improve a bit the help as shown by newshell. In particular, this groups sub-commands automatically so, when you do `w?` you will see `wv[1248]` because `wv` is a command, but you also have `wv1`, `wv2`, etc. If there are too many sub-commands, you see `wv[?]`. Also, if `wv` by itself is not a command, you would see `wv<1248>`, to make sure a user can tell the difference between a command that can work by itself and one that cannot.

Below you can see some screens of the current state. (Please note that `w<1248><+->` is not automatically generated :(, but I let commands define their own list of options, when that's required in some cases).
image
image
image
image
image

### CODE REVIEW

File: libr/core/cmd_api.c [hunk 1]
```diff
@@ -1162,6 +1235,19 @@ R_API RCmdDesc *r_cmd_desc_parent(RCmdDesc *cd) {
 	return cd->parent;
 }
 
+R_API bool r_cmd_desc_has_handler(RCmdDesc *cd) {
+	r_return_val_if_fail (cd, NULL);
```
Comment (trufae) [inline libr/core/cmd_api.c:1239]: false not null

File: libr/core/cmd_api.c [hunk 2]
```diff
@@ -1162,6 +1235,19 @@ R_API RCmdDesc *r_cmd_desc_parent(RCmdDesc *cd) {
 	return cd->parent;
 }
 
+R_API bool r_cmd_desc_has_handler(RCmdDesc *cd) {
```
Comment (trufae) [inline libr/core/cmd_api.c:1238]: this function only can return true if the .cb is set on argv mode, all the other cases is false, can you simplify this logic to just handle this?

File: libr/core/cmd_api.c [hunk 3]
```diff
@@ -418,29 +420,95 @@ static void fill_usage_strbuf(RStrBuf *sb, RCmdDesc *cd, bool use_color) {
...
+static bool show_group_args(RCmdDesc *cd, RCmdDesc *parent) {
+	return !show_args (cd, parent) && cd->help->group_args_str;
+}
+
+static size_t calc_padding_len(RCmdDesc *cd, RCmdDesc *parent) {
+	size_t name_len = strlen (cd->name);
+	size_t args_len = 0;
+	size_t children_length = 0;
+	if (show_children_shortcut (cd, parent)) {
+		char *children_s = children_chars (cd);
```
Comment (trufae) [inline libr/core/cmd_api.c:471]: check for null

File: libr/core/cmd_api.c [hunk 4]
```diff
@@ -418,29 +420,95 @@ static void fill_usage_strbuf(RStrBuf *sb, RCmdDesc *cd, bool use_color) {
...
+		   *pal_opt_color = use_color? cons->context->pal.reset: "",
 		   *pal_help_color = use_color? cons->context->pal.help: "",
 		   *pal_input_color = use_color? cons->context->pal.input: "",
 		   *pal_reset = use_color? cons->context->pal.reset: "";
 
-	r_strbuf_appendf (sb, "| %s%s%s%s %*s%s# %s%s\n", pal_input_color, cd->name,
-		pal_args_color, cd_args_str, padding, "", pal_help_color, cd_summary, pal_reset);
+	r_strbuf_appendf (sb, "| %s%s", pal_input_color, cd->name);
+	if (show_children_shortcut (cd, parent)) {
+		char *children_s = children_chars (cd);
```
Comment (trufae) [inline libr/core/cmd_api.c:502]: check for null

**Review by trufae — CHANGES_REQUESTED** [review]

File: libr/core/cmd_api.c (hunk 2 above)
Comment (ret2libc) [reply libr/core/cmd_api.c:1238]: No there is also the old input case that can return true

**Review by XVilka — APPROVED** [review]
Good idea.

File: test/unit/test_cmd.c [hunk 5]
```diff
@@ -288,8 +288,8 @@ bool test_cmd_help(void) {
 	r_cmd_desc_oldinput_new (cmd, p_cd, "px", px_handler, &px_help);
 
 	const char *p_help_exp = "Usage: p-usage   # p summary\n"
-		"| pd <num>                 # pd summary\n"
-		"| px <verylongarg_str_num> # px summary\n";
+		"| pd <num>                    # pd summary\n"
+		"| px[?] <verylongarg_str_num> # px summary\n";
```
Comment (trufae) [inline test/unit/test_cmd.c:292]: can this verylongarg string be shorter?

File: libr/core/cmd_api.c [hunk 6]
```diff
@@ -7,6 +7,8 @@
 #include <r_cmd.h>
 #include <r_util.h>
 
+#define MAX_CHILDREN_SHOW 7
```
Comment (trufae) [inline libr/core/cmd_api.c:10]: why seven? can you add a comment explaining the use of this?

**Review by trufae — CHANGES_REQUESTED** [review]

File: test/unit/test_cmd.c (hunk 5 above)
Comment (ret2libc) [reply test/unit/test_cmd.c:292]: what do you mean? this is a test to make sure the summaries (the text on the right) are aligned. What's the problem with having a long string in the tests??

File: libr/core/cmd_api.c (hunk 6 above)
Comment (ret2libc) [reply libr/core/cmd_api.c:10]: Random number... I had to set a limit somewhere. I will add a comment.

**Review by trufae — APPROVED** [review]

File: test/unit/test_cmd.c (hunk 5 above)
Comment (trufae) [reply test/unit/test_cmd.c:292]: oh nvmits just a unit test

**Comment by trufae** [comment]
make it grin

**Pull request merged**